lib/orders/use_case/add_beverage_to_order.rb

```diff
@@ -0,0 +1,34 @@
+module Orders
+  module UseCase
+    class AddBeverageToOrder
+      Request = Struct.new(:order_id, :beverage)
+
+      attr_reader :gateway
+
+      def initialize(gateway: Gateway::Order.new)
+        @gateway = gateway
+      end
+
+      def add_beverage_to_order(request)
+        unless request[:beverage].class.ancestors.include?(Orders::Domain::Beverage)
```

Comment:
I'd avoid doing type checking. 
In Ruby, if it quacks like a duck then you just get on with it. 

The error you'll get is a runtime error along the lines of `method missing` or `invalid arguments`

_As an aside - This is the same as `instanceof` in C#. In a statically-typed context you should only do that if this class should "know" the type already. E.g. if you have a class that only deals with a certain subclass, and fetches a record that it knows is a specific subclass (because it requested it), then a cast is allowed in C#/Java._

Reply:
A wider point here is that Controllers should not know about Domain objects, so to receive a domain object in the request would break that rule.

The "boundary" between Controllers/Delivery Mechanism and Use Cases should consist of pure data structures only.

![image](https://user-images.githubusercontent.com/1889973/182425534-becb7e08-b454-49e9-b9fa-34416ec9df1a.png)